docs/api/operators.md

```diff
@@ -0,0 +1,248 @@
+);
+
+// Prints 56789 (1+4, 2+4, 3+4, 3+5, 3+6) to the console.
+```
+
+## concat
+
+`concat` will combine two sources together, subscribing to the next source after the previous source completes.
+
+```reason
+open Wonka;
+open Wonka_sources;
+open Wonka_operators;
+open Wonka_sinks;
+
+let sourceOne = fromArray([|1, 2, 3, 4, 5, 6|]);
+let sourceTwo = fromArray([|6, 5, 4, 3, 2, 1|]);
+
+concat([|sourceOne, sourceTwo|]) |> subscribe((. _val) => print_int(_val));
+
+/* Prints 1 2 3 4 5 6 6 5 4 3 2 1 to the console. */
+```
+
+```typescript
+import { fromArray, pipe, concat, subscribe } from 'wonka';
+
+const sourceOne = fromArray([1, 2, 3, 4, 5, 6]);
+const sourceTwo = fromArray([6, 5, 4, 3, 2, 1]);
+
+pipe(
+  concat([sourceOne, sourceTwo]),
+  subscribe(val => {
+    console.log(val);
+  })
+);
+
+// Prints 1 2 3 4 5 6 6 5 4 3 2 1 to the console.
+```
+
+## filter
+
+`filter` will remove values from a source by passing them through an iteratee that returns a `bool`.
+
+```reason
+open Wonka;
+open Wonka_sources;
+open Wonka_operators;
+open Wonka_sinks;
+
+let source = fromArray([|1, 2, 3, 4, 5, 6|]);
+let isEven = (. n) => n mod 2 === 0;
+
+source |> filter(isEven) |> subscribe((. _val) => print_int(_val));
+
+/* Prints 246 to the console. */
+```
+
+```typescript
+import { fromArray, filter, subscribe } from 'wonka';
+
+const source = fromArray([1, 2, 3, 4, 5, 6]);
+const isEven = n => n % 2 === 0;
+
+pipe(
+  source,
+  filter(isEven),
+  subscribe(val => {
+    console.log(val);
+  })
+);
+
+// Prints 246 to the console.
+```
+
+## map
+
+`map` will transform values from a source by passing them through an iteratee that returns a new value.
+
+```reason
+open Wonka;
+open Wonka_sources;
+open Wonka_operators;
+open Wonka_sinks;
+
+let source = fromArray([|1, 2, 3, 4, 5, 6|]);
+let square = (. n) => n * n;
+
+source |> map(square) |> subscribe((. _val) => print_int(_val));
+
+/* Prints 1 4 9 16 25 36 to the console. */
+```
+
+```typescript
+import { fromArray, pipe, map, subscribe } from 'wonka';
+
+const source = fromArray([1, 2, 3, 4, 5, 6]);
+const square = n => n * n;
+
+pipe(
+  source,
+  map(square),
+  subscribe(val => {
+    console.log(val);
+  })
+);
+
+// Prints 1 4 9 16 25 36 to the console.
+```
+
+## merge
+
+`merge` two sources together into a single source.
+
+```reason
+open Wonka;
+open Wonka_sources;
+open Wonka_operators;
+open Wonka_sinks;
+
+let sourceA = fromArray([|1, 2, 3|]);
+let sourceB = fromArray([|4, 5, 6|]);
+
+merge([|sourceA, sourceB|]) |> subscribe((. _val) => print_int(_val));
+
+/* Prints 1 2 3 4 5 6 to the console.
+```
+
+```typescript
+import { fromArray, pipe, merge, subscribe } from 'wonka';
+
+const sourceOne = fromArray([1, 2, 3]);
+const sourceTwo = fromArray([4, 5, 6]);
+
+pipe(
+  merge(sourceOne, sourceTwo)
+  subscribe((val) => {
+    console.log(val);
+  })
+);
+
+// Prints 1 2 3 4 5 6 to the console.
+```
+
+## onEnd
+
+Run a callback _after_ all values have been emitted by the source.
```

Comment:
I'm not 100% sure this is an accurate definition. I wrote a similar example with `onEnd` and got the same output logged to the console, which was confusing 🤔 Would you mind chiming in on the examples for `onEnd`, `onStart`, and `onPush`? I didn't see test cases for them either.

Reply:
All these operators merely represent the internal events, or rather signals, that are sent to the sink.

So onEnd fires when it receives End, onPush when it receives Push, and onStart when it receives Start.

So technically this is again only a relevant operator for asynchronous streams. It does fire after all values have been sent in the synchronous case as that’s when the source synchronously ends. So that’s 100% equivalent to it being “out of values” so after all of them have been sent, which, again, is very predictable.

In an asynchronous case the End event can be delayed. There’s no rule saying that you need to end the source when no value is going to be sent anymore. Think of a web socket connection suddenly becoming unstable and being closed by the browser. In such a case the End event wouldn’t correspond with the incoming values at all